django/db/backends/postgresql_psycopg2/client.py

```diff
@@ -1,19 +1,66 @@
+from __future__ import print_function
+
+import os
 import subprocess
+import tempfile
 
 from django.db.backends.base.client import BaseDatabaseClient
 
 
+def _escape_pgpass(txt):
+    """
+    Escape a non-terminal fragment of a postgresql .pgpass file.
+    """
+    return txt.replace('\\', '\\\\').replace(':', '\\:')
+
+
 class DatabaseClient(BaseDatabaseClient):
     executable_name = 'psql'
 
     def runshell(self):
         settings_dict = self.connection.settings_dict
         args = [self.executable_name]
-        if settings_dict['USER']:
-            args += ["-U", settings_dict['USER']]
-        if settings_dict['HOST']:
-            args.extend(["-h", settings_dict['HOST']])
-        if settings_dict['PORT']:
-            args.extend(["-p", str(settings_dict['PORT'])])
-        args += [settings_dict['NAME']]
-        subprocess.call(args)
+
+        host = settings_dict.get('HOST', '')
+        port = settings_dict.get('PORT', '')
+        name = settings_dict.get('NAME', '')
+        user = settings_dict.get('USER', '')
+        passwd = settings_dict.get('PASSWORD', '')
+
+        if user:
+            args += ["-U", user]
+        if host:
+            args += ["-h", host]
+        if port:
+            args += ["-p", str(port)]
+        args += [name]
+
+        temp_pgpass = None
+        try:
+            if passwd:
+                # Create temporary .pgpass file. This requires PostgreSQL >= 8.1.
+                temp_pgpass = tempfile.NamedTemporaryFile(mode='w+', delete=False)
+
+                # Note that print supports non ascii7 characters in strings on
+                # python 2 and 3, and both on systems with utf8 and other locales.
+                print(
+                    _escape_pgpass(host) or '*',
+                    str(port) or '*',
+                    _escape_pgpass(name) or '*',
+                    _escape_pgpass(user) or '*',
+                    passwd,
+                    file=temp_pgpass,
+                    sep=':',
+                )
+
+                # We need to close the file and remove it manually because
```

Comment:
Did you have a look if `django.core.files.temp.NamedTemporaryFile` might help here?